### DISCUSSION TITLE
Allow users to change the theme

### DISCUSSION THREAD

**Discussion by benthecoder** [osmoscraft/osmosfeed#16] [Ideas] [answered]
Really love this! Can't wait to see more development and features! I'm sorry if this is a bad question, but is there a way to choose templates/themes for the rss reader? Or is this theme the only one that's available now?

**Comment by benthecoder** [comment]
Another question I have is whether the rss feeds can be centered because I use a big monitor and all the feeds are at the left which isn't a nice viewing experience. I would rather have everything at the center. Is this possible? Thank you.

**Comment by spitlo** [comment]
If there was a way to include your own stylesheet, that would be great! Either by supplying a path to a custom stylesheet in `osmosfeed.yaml`, or by simply placing a `styles.css` file in the root.

**Comment by chuanqisun** [comment]
Thank you for bringing this up @benthecoder and thank you @spitlo for the idea. I definitely plan to support theming and that will be a core feature of this tool. The current theme ([gruvbox](https://github.com/morhetz/gruvbox)) is implemented with the [base16 architecture](https://github.com/chriskempson/base16) so swapping shouldn't be hard.

Please note, this project suddenly received way more attention than I expected, so please give me some time to get the technical design and implementation right for this very important feature. I want to keep the system modular so the basic experience is always simple and lightweight, but for advanced users, they will have full control, either through `osmosfeed.yaml`, or extensions/plugins via npm.

Thank you again.

**Comment by chuanqisun** [comment, marked as answer]
Good news, with #21, light theme (and any base16 theme) is now possible. See [Demo](https://osmoscraft.github.io/osmosfeed-examples/default-gruvbox-light/) | [Source](https://github.com/osmoscraft/osmosfeed-examples/tree/main/examples/default-gruvbox-light). You can try other themes with this tool: https://terminal.sexy/

The default template exposes all theme variables in the `:root {}` css declaration. The example shows how you can change the parameters. For advanced theming, you can write additional css in that file. Let me know if this works for you @benthecoder, @spitlo.

Technical note: I chose to inline custom snippet into the final output because GitHub host has a 10 minute cache on linked file. I don't want to deal with cache invalidation at this point.

**Reply by chuanqisun** [reply]
![image](https://user-images.githubusercontent.com/1895289/117109492-9afdd680-ad39-11eb-944e-db10be47c83f.png)

**Reply by spitlo** [reply]
That’s great, thanks a lot!

**Reply by onnyyonn** [reply]
It seems not all base16 variables are being used in the theme. I did some testing and it seems that only the following variables are in use:
base00 - background 1
base01 - background 2
base02 - background 3
base04 - Name of the feed, Read button, Footer (timestamp of last update, osmosfeed version)
base06 - Title, Text under a title
base0A - Date
base0D - Hovered url, Hovered title

Maybe some more variables can be used to give more granular control? A documentation of which variable corresponds to what component will be great too.

**Reply by chuanqisun** [reply]
Thank you @onnyyonn for giving the theming system a try. I hope [release v1.6.0](https://github.com/osmoscraft/osmosfeed/releases/tag/v1.6.0) has addressed your questions.
- The latest css takes base16 values and map them into intermediary variables.
- You can adjust those intermediary variables to have fine control on the look and feel of the UI
- I adopted [BEM](http://getbem.com/naming/) so you can select most elements with ease and style them with arbitrary CSS.
- I added [more documentation](https://github.com/osmoscraft/osmosfeed/blob/master/docs/customization-guide.md#theming-guide)
- I added [more examples](https://github.com/osmoscraft/osmosfeed-examples/tree/main/examples)

![image](https://user-images.githubusercontent.com/1895289/117631879-881c4500-b131-11eb-9519-96056d670075.png)

As a side note, I hope to ship the custom templating system soon so you can have full control over what html to work with. Until then, people will need to work with the default template and the gruvbox theme, so I hope to keep the color usage from the theme at a minimum, just to make a cleaner starting point for everyone.

Let me know if you need additional help. Would love to see how you've customized it!

**Reply by onnyyonn** [reply]
That's great news! I am excited about the upcoming templating engine! For now I have reverted back to 1.5.0, because the new update was breaking my hacky theme. You can check it out here https://onnyyonn.github.io/feed . Once the template API is available, I'll try to make it a proper theme.

**Reply by chuanqisun** [reply]
That's sleek! I wonder, if I ever add a section somewhere in the docs to showcase cool buildouts from users, could I feature your work?

**Reply by onnyyonn** [reply]
Yeah please feel free to add it :)

**Reply by chuanqisun** [reply]
Fully customizable templating system is here with release [v1.7.0](https://github.com/osmoscraft/osmosfeed/releases/tag/v1.7.0). I adopted [Handlebar](https://handlebarsjs.com/guide/) as the templating language.

- [Demo](https://osmoscraft.github.io/osmosfeed-examples/articles-unstyled/) ⚠ WARNING: it is going to hurt your eyes as I didn't apply any css
- [Source](https://github.com/osmoscraft/osmosfeed-examples/blob/main/examples/articles-unstyled/includes/index.hbs) 
- [Docs](https://github.com/osmoscraft/osmosfeed/blob/master/docs/customization-guide.md#template-customization-guide)

Thank you @onnyyonn  for motivating me into getting this feature worked out. Please let me know if meets your customization requirements.

**Reply by onnyyonn** [reply]
@chuanqisun Thank you. It works really well for all the customization I want to do. I have now updated my theme with a Handlebar template.
